### ISSUE TITLE
warning: using pulumi-resource-azure-native / pulumi-language-nodejs from $PATH

### ISSUE THREAD

**Issue by teto** [pulumi/pulumi#14525]
### What happened?

my pulumi commands run fine but are polluted with this warning about running software from PATH.
I could find little info about this (just https://github.com/pulumi/pulumi/issues/13656) and I wonder since when running software from PATH is a bad thing ? 
alternatively is there a way to silence just this warning ?

### Example

```
  pulumi:pulumi:Stack (ci-azure-runners-dev):
    warning: using pulumi-language-nodejs from $PATH at /nix/store/my9bsqgm1lb79i05zw8pcmc2i852qkh0-build-env/bin/pulumi-language-nodejs
    warning: using pulumi-language-nodejs from $PATH at /nix/store/my9bsqgm1lb79i05zw8pcmc2i852qkh0-build-env/bin/pulumi-language-nodejs
    warning: using pulumi-resource-azure-native from $PATH at /nix/store/my9bsqgm1lb79i05zw8pcmc2i852qkh0-build-env/bin/pulumi-resource-azure-native
    warning: using pulumi-resource-azure-native from $PATH at /nix/store/my9bsqgm1lb79i05zw8pcmc2i852qkh0-build-env/bin/pulumi-resource-azure-native
    warning: using pulumi-resource-azure-native from $PATH at /nix/store/my9bsqgm1lb79i05zw8pcmc2i852qkh0-build-env/bin/pulumi-resource-azure-native
    Found IP: [object Object]
```

### Output of `pulumi about`

CLI          
Version      3.90.1
Go Version   go1.21.3
Go Compiler  gc

Host     
OS       nixos
Version  23.11 (Tapir)
Arch     x86_64

Backend        
Name           jedha
URL            s3://redacted
User           redacted
Organizations  
Token type     personal

Pulumi locates its logs in /tmp/nix-shell.4baXP2 by default
warning: Failed to read project: no Pulumi.yaml project file found (searching upwards from /home/teto/nova/ci-runner). If you have not created a project yet, use `pulumi new` to do so: no project file found
warning: Failed to get information about the current stack: no Pulumi.yaml project file found (searching upwards from /home/teto/nova/ci-runner). If you have not created a project yet, use `pulumi new` to do so: no project file found

**Comment by Frassle** [comment]
In normal setups running these binaries from PATH is nearly always indicative of a mistake.

Even during development where we do frequently build the plugins to ~/go/bin (on PATH) and run them from there it is _incredibly_ useful to have the CLI remind you that it's running from PATH and not running the proper versioned plugin from the plugin cache.
Before this warning was added untold hours were wasted by our developers testing things and then noticing that either their local build hadn't copied to PATH and they were just testing the last released plugin, or confused why something that was supposed to be fixed didn't work in the latest release only to realize they were running against an old local build of the plugin still left on PATH.

Pulumi (the company) have basically no involvement in the packaging of pulumi for nix, so its not something we can easily fix but I'd argue that these things are not being packaged in correct way that the rest of the pulumi ecosystem expects.
`pulumi-language-nodejs` should be installed with `pulumi` into the same directory so we can tell it's bundled, and `pulumi-resource-azure-native` should be in pulumi's versioned plugin cache (~/.pulumi/plugins) so that the engine can handle multiple versions of the same plugin.

> alternatively is there a way to silence just this warning ?

No, and I'm not overly keen on adding that to fix arguably a bug in nix's packaging. Further adding a way to silence this warning risks devs turning that on and losing out on the whole point of why this was added.

**Comment by jjhuff** [comment]
> `pulumi-language-nodejs` should be installed with `pulumi `into the same directory so we can tell it's bundled, and `pulumi-resource-azure-native` should be in pulumi's versioned plugin cache (~/.pulumi/plugins) so that the engine can handle multiple versions of the same plugin.

This would appear to be the case for me:
```
$ whereis pulumi
pulumi: /nix/store/v7qgln7y0ag3yq17c2rdpda6r392gr44-devenv-profile/bin/pulumi
$ whereis pulumi-language-nodejs
pulumi-language-nodejs: /nix/store/v7qgln7y0ag3yq17c2rdpda6r392gr44-devenv-profile/bin/pulumi-language-nodejs
```

**Comment by Frassle** [comment]
Ah the symlinks here are complicating things. We've got a number of other official setups (like homebrew) using symlinks in some shape that the plugin lookup code is already having to deal with. It _might_ be possible to update the lookup code to handle the warning cases we want while correctly treating this nix setup as valid.

I am open to a PR to fix this as long as all current tests continue to pass. You'll want to look at the logic in getPluginInfoAndPath in sdk/go/common/workspace/plugins.go.

**Comment by jjhuff** [comment]
FWIW, I totally understand and support the need for a check. I've also wasted countless hours on similar issues!

I'm not sure there even IS a good way to fix this based on path alone. Solutions that come to mind (on the pulumi side):
* Bake a version into the plugin and check it by executing the plugin
* Baking it into the file name (a la `pulumi-language-nodejs-3.93.0`)
* https://github.com/pulumi/pulumi/issues/956 and treat language plugins like the resource plugins

On the nix side, there's another pulumi package (`pulumi-bin`) that includes everything. Unfortunately, it also includes all of the language plugins and thus causes problems there. I'm not super familiar with nix packaging, but I'll see if the maintainers have any other ideas.

Saving that, I can just ignore the warning :)

**Comment by thatsmydoing** [comment]
> and `pulumi-resource-azure-native` should be in pulumi's versioned plugin cache (~/.pulumi/plugins) so that the engine can handle multiple versions of the same plugin.

I don't know if it's still the case, but we had to install pulumi resource plugins via nix instead of using pulumi as the pre-compiled binaries wouldn't run on nixos without being patched. I suppose now you can build your own plugins and `pulumi plugin install --server` but that's a lot more inconvenient than using nix to install them and have them on $PATH.

It is quite non-standard but at least for us it works. I understand why you'd not want to add a flag to disable the warning but it would be appreciated.

**Comment by Frassle** [comment]
> but we had to install pulumi resource plugins via nix instead of using pulumi as the pre-compiled binaries wouldn't run on nixos without being patched

First I've heard of that, what exactly was the issue with the pre-compiled binaries that Pulumi releases?

**Comment by thatsmydoing** [comment]
It's not particular to pulumi but is a thing in general for NixOS (and other non-FHS distros like Guix and GoboLinux) https://nixos.wiki/wiki/Packaging/Binaries

IIUC the problem was that the linker is not in the "standard" path and the path has to be replaced with the correct one. I think this is only a problem if a plugin was built with CGO enabled. It seems that's no longer the case and the precompiled plugins are now static executables so run fine without patching.

We ran into the problem back in May 2020 when we were first setting up pulumi and it seems things were fixed pretty soon after but we haven't really updated our setup since then.

**Comment by Frassle** [comment]
Yeh we've moved away from CGO as much as possible, everything should be statically linked now.

**Comment by thatsmydoing** [comment]
Right, so that aside, from your previous comment, it seems the correct way to package pulumi is that the pulumi binary + bundled plugins should be all in the same directory and they should be in `$PATH`. It might be possible to fake being in the same directory by symlinking them into a single directory.

As for non-bundled plugins, having them be in `$PATH` is unsupported and users should install them using pulumi directly? I feel like there's some argument for allowing non-pulumi installed plugins which is why I think it might still make sense to be able to disable the warning.

**Comment by teto** [comment]
in my experience, these kinds of forced checks dont survive in the wild and so forcing the warning on the user without recourse is hard on the user. I dont see any solution to remove the warning in a nix env apart from patching the sourc.
More and more programs tend to have `health` or `doctor` command to check an install (I think about `:checkhealth` in neovim for instance or `nix doctor` ). It's a bit sad we needer the latter pattern but at least `pulumi doctor` would be on demand. Or maybe it does exist already.

**Comment by Frassle** [comment]
> these kinds of forced checks dont survive in the wild and so forcing the warning on the user without recourse is hard on the user

Understood, but firstly this isn't really a supported setup (nix) so I feel OK with it being a bit loud warning about things. Secondly, I'll requote my first comment:

> Even during development where we do frequently build the plugins to ~/go/bin (on PATH) and run them from there it is incredibly useful to have the CLI remind you that it's running from PATH and not running the proper versioned plugin from the plugin cache.

It is _really_ useful for our and third-party engineering to have this warning. The amount of times people used to get tripped up by out of date plugins on path was a lot, and since the warning we hardly see this mistake anymore. I _really_ don't want to have a way to turn this warning off because it defeats the whole point of it.

A `pulumi health` command would not have the same effect because the whole point is people think they're in a good state and just try and run updates then see the warning.

I remain open to a PR to fix the warning for this specific case of nix symlinks being a bit odd, as long as the warning system continues to work for the common case of people just leaving things on PATH.

**Comment by Frassle** [comment]
> can't we just remove the checks about symlinks ?

As long as it doesn't break homebrew setups which also heavily use symlinks (and is our supported distribution mechanism). Looking at the plugin code it might just be a case of needing to do some more symlink evals.

**Comment by voidus** [comment]
I think for now, something like this is an acceptable workaround

```nix
(runCommand "${pulumi.name}-with-packages" { } ''
  mkdir -p $out/bin
  cp ${pulumi}/bin/* $out/bin/
  cp ${pulumiPackages.pulumi-language-python}/bin/* $out/bin/
'')
```
